### PR TITLE
Issue #69: Refactor CoolUtils

### PR DESCRIPTION
**Pull request by donquixote**

### CODE REVIEW

File: src/MediaHelper.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+namespace Drupal\collabora_online;
+
+use Drupal\file\Entity\File;
+use Drupal\media\Entity\Media;
+
+/**
+ * Service to assist with media entities.
+ */
+class MediaHelper {
```
Comment (donquixote) [inline src/MediaHelper.php:13]: I can add interfaces for these classes, but we should first agree on the general direction and naming.

File: src/Controller/WopiController.php [hunk 2]
```diff
@@ -278,4 +313,42 @@ public function wopi(string $action, string $id, Request $request) {
...
+   *   The token to verify.
+   * @param int|string $expected_media_id
+   *   Media id expected to be in the token payload.
+   *   This could be a stringified integer like '123'.
+   *
+   * @return array|null
+   *   Data decoded from the token, or NULL on failure or if the token has
+   *   expired.
+   */
+  protected function verifyTokenForMediaId(
```
Comment (AaronGilMartinez) [inline src/Controller/WopiController.php:333]: Im thinking since we are decoding and verifing in the token manager keep the verify methods in the same place.
For me seems strange to have the same operation in the service and controller.

File: src/Cool/CoolUtils.php
Comment (AaronGilMartinez) [inline src/Cool/CoolUtils.php:1]: The method that is left (canEdit) misses the function return type (: bool) and it's using File instead of FileInterface, also strict_types. 

This is a bit on the limbo, we have a helper for media, and the URL helper to get the URLs to edit and view. I'm thiking that this last method may fit in media helper since we have different methods to do stuff with medias and files related to medias. What do you think?

File: src/CoolUrl.php
Comment (AaronGilMartinez) [inline src/CoolUrl.php:1]: This a Cool thing but it's not in the cool namespace, should we move it?

**Review by AaronGilMartinez — CHANGES_REQUESTED** [review]

File: src/Controller/ViewerController.php [hunk 3]
```diff
@@ -92,7 +95,20 @@ public function editor(MediaInterface $media, Request $request, $edit = FALSE) {
       );
     }
 
-    $render_array = CoolUtils::getViewerRender($media, $wopi_client_url, $edit, $options);
+    try {
+      $render_array = $this->getViewerRender($media, $wopi_client_url, $edit, $options);
+    }
+    catch (CollaboraNotAvailableException $e) {
+      $this->getLogger('cool')->warning(
+        "Cannot show the viewer/editor.<br>\n" . Error::DEFAULT_ERROR_MESSAGE,
```
[the remark this answers is not in the document]
Comment (donquixote) [reply src/Controller/ViewerController.php:103]: Do we call it a "viewer" if it is in edit mode?

File: src/Controller/ViewerController.php [hunk 4]
```diff
@@ -103,4 +119,57 @@ public function editor(MediaInterface $media, Request $request, $edit = FALSE) {
...
+   *   Options for the renderer. Current values:
+   *     - "closebutton" if "true" will add a close box. (see COOL SDK)
+   *
+   * @return array|array{error: string}
+   *   A stub render element array, or an array with an error on failure.
+   *
+   * @throws \Drupal\collabora_online\Exception\CollaboraNotAvailableException
+   *   The key to use by Collabora is empty or not configured.
+   */
+  protected function getViewerRender(MediaInterface $media, string $wopi_client, bool $can_write, $options = NULL) {
```
[the remark this answers is not in the document]
Comment (donquixote) [reply src/Controller/ViewerController.php:142]: Actually, in the current code there is so much wrong with these `$options`:
- The value is always the same (it is only called once).
- The value is `['closebutton' => 'true']`, using string `'true'` instead of boolean `true`.
- The array only has this one key.

So for now the simplest is to drop this parameter altogether.

File: src/Controller/ViewerController.php [hunk 5]
```diff
@@ -103,4 +119,57 @@ public function editor(MediaInterface $media, Request $request, $edit = FALSE) {
...
+    $render_array = [
+      '#wopiClient' => $wopi_client,
+      '#wopiSrc' => urlencode($wopi_base . '/cool/wopi/files/' . $id),
+      '#accessToken' => $access_token,
+      // Convert to milliseconds.
+      '#accessTokenTtl' => $expire_timestamp * 1000,
+      '#allowfullscreen' => $allowfullscreen ? 'allowfullscreen' : '',
+    ];
+    if ($options) {
+      if (isset($options['closebutton']) && $options['closebutton'] == 'true') {
```
[the remark this answers is not in the document]
Comment (donquixote) [reply src/Controller/ViewerController.php:167]: This code is mostly unchanged since it was moved.
I would like to rethink the full render element and the template, instead of tinkering around the edges.
This should be a follow-up issues.

This said, in this issue we can do _some_ cleanup.

File: src/Controller/WopiController.php [hunk 6]
```diff
@@ -26,6 +33,34 @@
...
+   * @param \Drupal\Core\Session\AccountSwitcherInterface $accountSwitcher
+   *   Account switcher.
+   * @param \Drupal\Core\File\FileSystemInterface $fileSystem
+   *   File system.
+   * @param \Drupal\Component\Datetime\TimeInterface $time
+   *   Time.
+   * @param \Drupal\Core\File\FileUrlGeneratorInterface $fileUrlGenerator
+   *   File url generator.
+   * @param \Drupal\collabora_online\MediaHelper $mediaHelper
+   *   Media helper.
```
[the remark this answers is not in the document]
Comment (donquixote) [reply src/Controller/WopiController.php:52]: True, core puts "The " in front of these.

Nowadays there is also this:
https://www.drupal.org/docs/develop/standards/php/api-documentation-and-comment-standards#classes

> All classes and all of their methods (including private methods but **excluding constructors**) must be documented.

But let's not go there for now.

File: src/MediaHelper.php [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  /**
+   * Sets the file entity reference for a media entity.
+   *
+   * @param \Drupal\media\Entity\Media $media
+   *   Media entity to be modified.
+   * @param \Drupal\file\Entity\File $source
+   *   File entity to reference.
+   */
+  public function setMediaSource(Media $media, File $source) {
+    $name = $media->getSource()->getSourceFieldDefinition($media->bundle->entity)->getName();
```
Comment (donquixote) [inline src/MediaHelper.php:55]: Btw @AaronGilMartinez This one is also dirty, but I don't think we can easily avoid it.

In `MediaType::sourcePluginCollection()` there is a `Drupal::service()` call:
```php
    if (!$this->sourcePluginCollection && $this->source) {
      $this->sourcePluginCollection = new DefaultSingleLazyPluginCollection(\Drupal::service('plugin.manager.media.source'), $this->source, $this->source_configuration);
    }
    return $this->sourcePluginCollection;
```

File: src/WopiTokenManager.php [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+use Drupal\key\KeyRepositoryInterface;
+use Firebase\JWT\JWT;
+use Firebase\JWT\Key;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\DependencyInjection\Attribute\Autowire;
+
+/**
+ * Service to manage the access token to be used by the WOPI client.
+ */
+class WopiTokenManager {
```
[the remark this answers is not in the document]
Comment (donquixote) [reply src/WopiTokenManager.php:29]: The JWT part is encapsulated into this service.
From the outside we don't care about it, we only care that it produces a key which it can later decode.
And this token is used for the WOPI requests. This is why I called it that.

[the remark this answers is not in the document]
Comment (donquixote) [reply src/WopiTokenManager.php:29]: So at least the interface does not need to have JWT in the name.

File: src/WopiTokenManager.php [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+   * Creates a JWT token for a media entity.
+   *
+   * The token will carry the following:
+   *
+   * - fid: the Media id in Drupal.
+   * - uid: the User id for the token. Permissions should be checked
+   *   whenever.
+   * - exp: the expiration time of the token.
+   * - wri: if true, then this token has write permissions.
+   *
```
[the remark this answers is not in the document]
Comment (donquixote) [reply src/WopiTokenManager.php:101]: True, this comment needed clean up after I changed how the method works.

File: src/WopiTokenManager.php [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+   * Obtains the signing key from the key storage.
+   *
+   * @return string
+   *   The key value.
+   *
+   * @throws \Drupal\collabora_online\Exception\CollaboraNotAvailableException
+   *   The key to use by Collabora is empty or not configured.
+   */
+  protected function getKey(): string {
+    $default_config = $this->configFactory->get('collabora_online.settings');
```
[the remark this answers is not in the document]
Comment (donquixote) [reply src/WopiTokenManager.php:151]: Yeah, I was wondering if we should introduce a dedicated service for this configuration.
I am hesitating because I don't find this elsewhere in Drupal.

Imo it is ok to do this kind of local refactoring first, and later we can unify the config handling.

File: src/WopiTokenManager.php [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+   *   The token to verify.
+   *
+   * @return array|null
+   *   Data decoded from the token, or NULL on failure or if the token has
+   *   expired.
+   *
+   * @throws \Drupal\collabora_online\Exception\CollaboraNotAvailableException
+   *   The key to use by Collabora is empty or not configured.
+   */
+  public function decodeAndVerify(
```
[the remark this answers is not in the document]
Comment (donquixote) [reply src/WopiTokenManager.php:69]: The clean thing would be to have two layers: One to decode/encode, another to set and then verify the timestamp.
I was going to create a Transcoder class, but I noticed that the JWT class is already that, except that it is static and throws many different exception types.

One difficulty is that the expire timestamp needs to be available outside of the decode() method.

File: src/Controller/ViewerController.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (AaronGilMartinez) [reply src/Controller/ViewerController.php:142]: Indeed

File: src/Controller/WopiController.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (AaronGilMartinez) [reply src/Controller/WopiController.php:52]: Yes, let's add only the article for the moment please.

File: src/WopiTokenManager.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (AaronGilMartinez) [reply src/WopiTokenManager.php:29]: But you are defining a service name based on how it's used rather than what it does. That's why I suggested Collabora, what about CollaboraJwtTokenManager and CollaboraTokenManagerInterface? We could remove jwt from class name if does not convince you.

File: src/Controller/ViewerController.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (AaronGilMartinez) [reply src/Controller/ViewerController.php:103]: Not totally sure about this, lets keep it as is for now.

File: src/Controller/WopiController.php (hunk 2 above)
Comment (donquixote) [reply src/Controller/WopiController.php:333]: The idea here is that the token manager should not care whether we are encoding media ids or something else.
We could change this design..

In fact we have 3 layers of functionality:
- The raw encode/decode.
- The timeout handling.
- The specific values relevant to a media entity.

This could give us different services each with their own interface..

File: src/Controller/ViewerController.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (donquixote) [reply src/Controller/ViewerController.php]: For now we always add the closebutton, in the future we can rethink it.

File: src/Controller/ViewerController.php [hunk 12]
```diff
@@ -103,4 +119,57 @@ public function editor(MediaInterface $media, Request $request, $edit = FALSE) {
...
+        'fid' => $id,
+        'uid' => $this->currentUser()->id(),
+        'wri' => $can_write,
+      ],
+      $expire_timestamp,
+    );
+
+    $render_array = [
+      '#wopiClient' => $wopi_client,
+      '#wopiSrc' => urlencode($wopi_base . '/cool/wopi/files/' . $id),
```
[the remark this answers is not in the document]
Comment (donquixote) [reply src/Controller/ViewerController.php:153]: The only unreliable part would be `$wopi_base`, which ideally we validate in the form.

Note that the `$wopi_base` is a url for the Drupal website, but it can be different from the public url of the Drupal website.

[the remark this answers is not in the document]
Comment (AaronGilMartinez) [reply src/Controller/ViewerController.php:153]: Yes, so enough for the moment.

File: src/Controller/ViewerController.php [hunk 13]
```diff
@@ -103,4 +115,50 @@ public function editor(MediaInterface $media, Request $request, $edit = FALSE) {
...
+      '#accessToken' => $access_token,
+      // Convert to milliseconds.
+      '#accessTokenTtl' => $expire_timestamp * 1000,
+      '#allowfullscreen' => $allowfullscreen ? 'allowfullscreen' : '',
+    ];
+    $render_array['#closebutton'] = 'true';
```
Comment (AaronGilMartinez) [inline src/Controller/ViewerController.php:159]: Why not setting the value in the variable initialization?

File: src/Controller/WopiController.php (hunk 2 above)
Comment (AaronGilMartinez) [reply src/Controller/WopiController.php:339]: Actually yes, makes sense to keep them separated, adding all those services is too much.

File: src/MediaHelper.php (hunk 7 above)
Comment (AaronGilMartinez) [reply src/MediaHelper.php:67]: Indeed, not sure if we can avoid.

File: src/CoolUrl.php
Comment (AaronGilMartinez) [reply src/CoolUrl.php:1]: Static is OK for this case. About naming, at the end is another helper, maybe UrlHelper is ok.

File: src/Cool/CoolUtils.php
Comment (AaronGilMartinez) [reply src/Cool/CoolUtils.php:1]: Well, since we are refactoring CoolUtils we should decide if its ok as is, or we move it.

Actually, I'd change the name, is not clear what it does. 
There a more things that can do a file editable, the name is a bit vague.
The method is to check if the mime type of a file is supported. What a bout: isFileTypeSupported

File: src/Controller/WopiController.php [hunk 14]
```diff
@@ -51,37 +86,37 @@ public static function permissionDenied(): Response {
...
     // @todo What if the uid in the payload is not set?
     // @todo What if $user is NULL?
-    $user = User::load($jwt_payload->uid);
-    $can_write = $jwt_payload->wri;
+    $user = User::load($jwt_payload['uid']);
+    $can_write = $jwt_payload['wri'];
 
     if ($can_write && !$media->access('edit in collabora', $user)) {
-      \Drupal::logger('cool')->error('Token and user permissions do not match.');
+      $this->getLogger('cool')->error('Token and user permissions do not match.');
```
[the remark this answers is not in the document]
Comment (donquixote) [reply src/Controller/WopiController.php]: changed it.

File: src/CoolUrl.php
Comment (donquixote) [reply src/CoolUrl.php:1]: For the record, here is a multiline regex to find classes with only static methods :)
Works in PhpStorm.
```
^class \w+ \{(?!(
.*)*
  +(public|protected|private) function)(
.*)*static function
```

Conclusion: Core has some such classes that are named *Helper, but more without this suffix.
It also has classes named *Helper that are not all-static.

Personally, a class name with *er as in SomethingDoer implies that we use it as an object.
But in the end it is a matter of convention, and expectations based mostly on familiarity.

Comment (donquixote) [reply src/CoolUrl.php:1]: Also, in core, many of these classes are in `Drupal\Component\Utility` namespace.

What about `Drupal\collabora_online\Utility\CollaboraUrl`?

Comment (donquixote) [reply src/CoolUrl.php:1]: Btw core does have a class `UrlHelper`.

Comment (donquixote) [reply src/CoolUrl.php:1]: I would say normally we don't need to have Collabora or Cool prefix in class names, there is already the namespace.
But we can add it if we already know other classes in core that are widely used that have the same name.

Comment (AaronGilMartinez) [reply src/CoolUrl.php:1]: CollaboraUrl is Ok

File: src/MediaHelperInterface.php [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+
+  /**
+   * Sets the file entity reference for a media entity.
+   *
+   * @param \Drupal\media\MediaInterface $media
+   *   Media entity to be modified.
+   * @param \Drupal\file\Entity\File $file
+   *   File entity to reference.
+   */
+  public function setMediaSource(MediaInterface $media, File $file): void;
```
Comment (AaronGilMartinez) [inline src/MediaHelperInterface.php:59]:
```suggestion
  public function setMediaSource(MediaInterface $media, FileInterface $file): void;
```

File: tests/src/Kernel/CollaboraMediaAccessTest.php [hunk 16]
```diff
@@ -18,58 +18,21 @@
...
-    'system',
-    'file',
-  ];
+class CollaboraMediaAccessTest extends CollaboraKernelTestBase {
 
   /**
    * {@inheritdoc}
    */
-  protected function setUp(): void {
+  public function setUp(): void {
```
Comment (AaronGilMartinez) [inline tests/src/Kernel/CollaboraMediaAccessTest.php:31]: Why this change of visibility?

File: tests/src/Unit/JwtTranscoderBaseTest.php [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+      'a' => 'A',
+      'exp' => $expire_timestamp,
+    ];
+    $this->assertSame($expected_payload, $payload);
+    $this->assertSame(
+      $expected_payload,
+      (array) json_decode(JWT::urlsafeB64Decode(
+        explode('.', $token)[1],
+      )),
+      'The token body part contains the data.',
```
Comment (AaronGilMartinez) [inline tests/src/Unit/JwtTranscoderBaseTest.php:51]:
```suggestion
      'The token body part does not contains the data.',
```

File: tests/src/Unit/JwtTranscoderBaseTest.php [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+    $this->assertNull($payload);
+    $this->assertSame(
+      [
+        'a' => 'A',
+        'exp' => $expire_timestamp,
+      ],
+      (array) json_decode(JWT::urlsafeB64Decode(
+        explode('.', $token)[1],
+      )),
+      'The token body part contains the data.',
```
Comment (AaronGilMartinez) [inline tests/src/Unit/JwtTranscoderBaseTest.php:75]: Same here

**Review by AaronGilMartinez — CHANGES_REQUESTED** [review]

File: tests/src/Kernel/CollaboraMediaAccessTest.php (hunk 16 above)
Comment (donquixote) [reply tests/src/Kernel/CollaboraMediaAccessTest.php]: I removed this and then re-added manually with wrong visibility. Good catch!

File: tests/src/Unit/JwtTranscoderBaseTest.php (hunk 17 above)
Comment (donquixote) [reply tests/src/Unit/JwtTranscoderBaseTest.php:51]: But it does though.
What this assertion attempts to verify is that the transcoder really does JWT, and not simply serialize() and unserialize() or some other method.
An alternative (perhaps better) would be to simply test the literal value, which should be fixed.

File: tests/src/Unit/JwtTranscoderBaseTest.php (hunk 18 above)
Comment (donquixote) [reply tests/src/Unit/JwtTranscoderBaseTest.php:75]: It does contain the data.
Again, perhaps we should verify the literal token value instead.

File: tests/src/Unit/JwtTranscoderBaseTest.php (hunk 17 above)
Comment (donquixote) [reply tests/src/Unit/JwtTranscoderBaseTest.php]: changed it to "should contain".
See https://www.drupal.org/project/coding_standards/issues/3131946

**Review by AaronGilMartinez — APPROVED** [review]
